File: frontend/src/shared/api/shootings.ts
import { getApiBaseUrl } from "@/shared/api/client";

/** Metadata response from /shootings/meta endpoint */
export interface ShootingsMeta {
  /** Content-based version hash */
  version: string;
  /** ISO timestamp when data was generated */
  generated_at: string;
  /** Total number of rows in dataset */
  rows: number;
  /** Available years in the dataset */
  years: number[];
  /** URL to fetch versioned NDJSON rows */
  rows_url: string;
  /** URL to fetch versioned GeoJSON */
  geojson_url: string;
}

/** Result of fetching meta with conditional request support */
export interface MetaFetchResult {
  /** Whether the data was modified (false = 304 Not Modified) */
  modified: boolean;
  /** Meta data (only present if modified) */
  meta?: ShootingsMeta;
}

/**
 * Fetches shootings metadata with conditional request support (ETag/304).
 *
 * @param lastKnownVersion - Optional last known version for If-None-Match header
 * @returns Promise resolving to meta fetch result
 * @throws Error if the API request fails
 *
 * @example
 * ```ts
 * // First fetch (no version known)
 * const result = await fetchShootingsMeta();
 * if (result.modified) {
 *   console.log(`Version: ${result.meta?.version}`);
 * }
 *
 * // Subsequent fetch with version check
 * const result2 = await fetchShootingsMeta(lastVersion);
 * if (!result2.modified) {
 *   console.log("Data unchanged, use cached version");
 * }
 * ```
 */
export async function fetchShootingsMeta(
  lastKnownVersion?: string | null
): Promise<MetaFetchResult> {
  const baseUrl = getApiBaseUrl();
  const url = `${baseUrl}/shootings/meta`;

  const headers: HeadersInit = {};
  if (lastKnownVersion) {
    headers["If-None-Match"] = `"${lastKnownVersion}"`;
  }

  const response = await fetch(url, { headers });

  // 304 Not Modified - data unchanged
  if (response.status === 304) {
    return { modified: false };
  }

  if (!response.ok) {
    throw new Error(
      `API request failed: ${response.status} ${response.statusText}`
    );
  }

  const meta = (await response.json()) as ShootingsMeta;
  return { modified: true, meta };
}

/**
 * Fetches shootings rows as NDJSON and parses to array of row objects.
 *
 * @param rowsUrl - Relative URL to the versioned NDJSON endpoint (from meta.rows_url)
 * @returns Promise resolving to array of row objects
 * @throws Error if the API request fails
 *
 * @example
 * ```ts
 * const rows = await fetchShootingsRows("/shootings/rows/abc123def456.ndjson");
 * console.log(`Loaded ${rows.length} rows`);
 * ```
 */
export async function fetchShootingsRows(
  rowsUrl: string
): Promise<Record<string, unknown>[]> {
  const baseUrl = getApiBaseUrl();
  const url = `${baseUrl}${rowsUrl}`;

  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(
      `API request failed: ${response.status} ${response.statusText}`
    );
  }

  // Parse NDJSON: one JSON object per line
  const text = await response.text();
  const lines = text.trim().split("\n");

  return lines
    .filter((line) => line.length > 0)
    .map((line) => JSON.parse(line) as Record<string, unknown>);
}

File: packages/api/app/routers/shootings.py
"""Shootings endpoints.

This module provides endpoints for accessing shootings data:
- `/shootings/meta` - Primary entry point with version, metadata, and per-year URLs
- `/shootings/rows/{version}/{year}.ndjson` - Year-specific NDJSON for Arquero (immutable)
"""

import hashlib
import json
from collections.abc import Iterator
from typing import Any, cast

from fastapi import APIRouter, Depends, Header, HTTPException, Request, Response
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field

from app.data_loader import make_refresh_dependency

router = APIRouter(dependencies=[Depends(make_refresh_dependency(["shootings"]))])

# Cache headers for immutable versioned assets (1 year)
IMMUTABLE_CACHE_CONTROL = "public, max-age=31536000, immutable"
# Cache headers for meta endpoint (must revalidate)
META_CACHE_CONTROL = "max-age=0, must-revalidate"


class ShootingYearManifest(BaseModel):
    """Row count and versioned download URL for one calendar year."""

    rows: int = Field(description="Number of shooting-victim rows for the year.")
    rows_url: str = Field(description="Versioned NDJSON path for the year.")


class ShootingsManifest(BaseModel):
    """Current shooting-data version and its year-specific downloads."""

    version: str = Field(description="Content-derived version of the current dataset.")
    generated_at: str = Field(description="UTC time when this API manifest was generated.")
    rows: int = Field(description="Total number of shooting-victim rows.")
    years: list[int] = Field(description="Calendar years available for download.")
    years_meta: dict[int, ShootingYearManifest] = Field(
        description="Row count and NDJSON path for each available year."
    )


NDJSON_RESPONSE: dict[int | str, dict[str, Any]] = {
    200: {
        "description": (
            "Newline-delimited JSON with one processed shooting-victim record per line."
        ),
        "content": {
            "application/x-ndjson": {
                "schema": {
                    "type": "string",
                    "description": (
                        "An NDJSON stream. Each line is one JSON object representing one "
                        "shooting victim, including normalized date fields, coordinates "
                        "when available, published or normalized demographic fields, and "
                        "geographic joins."
                    ),
                }
            }
        },
    }
}


@router.get(
    "/shootings/meta",
    response_model=ShootingsManifest,
    responses={304: {"description": "The current manifest version has not changed."}},
)
def get_shootings_meta(
    request: Request,
    response: Response,
    if_none_match: str | None = Header(None),
) -> Response | dict[str, Any]:
    """Return metadata about the shootings dataset including version and URLs.

    This is the primary entry point. The frontend should:
    1. Fetch this endpoint to get the current version and available years
    2. Use `years_meta[year].rows_url` to fetch NDJSON data for a specific year

    Supports ETag-based caching: if `If-None-Match` header matches the
    current version, returns 304 Not Modified.

    Parameters
    ----------
    request : fastapi.Request
        The current request with access to application state.
    response : fastapi.Response
        The response object for setting headers.
    if_none_match : str | None
        Optional ETag from client for conditional request.

    Returns
    -------
    Response | dict[str, Any]
        304 Not Modified if ETag matches, otherwise the metadata dict.
    """
    meta = request.app.state.shootings_meta
    etag = hashlib.sha256(
        json.dumps(meta, sort_keys=True, separators=(",", ":")).encode("utf-8")
    ).hexdigest()
    headers = {
        "ETag": f'"{etag}"',
        "Cache-Control": META_CACHE_CONTROL,
    }

    # Check for conditional request
    if if_none_match and if_none_match.strip('"') == etag:
        return Response(status_code=304, headers=headers)

    # Set cache headers
    response.headers.update(headers)

    return cast(dict[str, Any], meta)


@router.get(
    "/shootings/rows/{version}/{year}.ndjson",
    response_class=StreamingResponse,
    responses=NDJSON_RESPONSE,
)
def get_shootings_rows_by_year(
    request: Request,
    version: str,
    year: int,
) -> StreamingResponse:
    """Return shootings data for a specific year as NDJSON.

    Each line is a JSON object representing a flattened shooting record with:
    - `lon`, `lat`: Coordinates extracted from geometry
    - `dateInMs`: Date as Unix timestamp in milliseconds
    - `timeInMs`: Milliseconds since midnight
    - `weekday`: Day of week (0=Sunday, 6=Saturday)
    - `year`: Year of incident
    - All properties from the original GeoJSON features

    This endpoint is versioned and immutable - the same version always returns
    the same data, enabling aggressive client-side caching.

    Parameters
    ----------
    request : fastapi.Request
        The current request with access to application state.
    version : str
        The version string (must match current dataset version).
    year : int
        The year to fetch data for.

    Returns
    -------
    StreamingResponse
        NDJSON stream with immutable cache headers.

    Raises
    ------
    HTTPException
        404 if version doesn't match or year not available.
    """
    current_version = request.app.state.shootings_version
    if version != current_version:
        raise HTTPException(
            status_code=404,
            detail=f"Version '{version}' not found. Current version is '{current_version}'.",
        )

    rows_by_year = request.app.state.shootings_rows_by_year
    if year not in rows_by_year:
        available_years = sorted(rows_by_year.keys())
        raise HTTPException(
            status_code=404,
            detail=f"Year {year} not found. Available years: {available_years}",
        )

    def generate_ndjson() -> Iterator[str]:
        """Stream NDJSON rows for the specified year."""
        for row in rows_by_year[year]:
            yield json.dumps(row, separators=(",", ":")) + "\n"

    return StreamingResponse(
        generate_ndjson(),
        media_type="application/x-ndjson; charset=utf-8",
        headers={
            "Cache-Control": IMMUTABLE_CACHE_CONTROL,
            "X-Robots-Tag": "noindex",
        },
    )
